File: src/main/java/org/algo4j/linear/Container.java
package org.algo4j.linear;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;

/**
 * Created by ice1000 on 2017/1/11.
 *
 * @author ice1000
 */
public interface Container<T> extends
		Cloneable,
		Serializable,
		Iterable<T> {
	@Contract(pure = true)
	int size();

	@Contract(pure = true)
	boolean empty();

	void push(@NotNull T obj);

	@Nullable
	@Contract(pure = true)
	T top();

	void pop();

	@Nullable
	default T peek() {
		T ret = top();
		pop();
		return ret;
	}

	@Contract(pure = true)
	Container<T> clone();
}

File: src/main/kotlin/org/algo4j/Container.kt
@file:JvmName("Extension")
@file:JvmMultifileClass

/**
 * Created by ice1000 on 2017/2/1.
 *
 * @author ice1000
 */
package org.algo4j

import org.algo4j.linear.Queue
import org.algo4j.linear.RollingArray
import org.algo4j.linear.Stack
import org.jetbrains.annotations.Contract


@Contract(pure = true)
operator fun <T> RollingArray<T>.get(index: Int) =
		get(index)

@Contract(pure = true)
operator fun <T> RollingArray<T>.set(index: Int, value: T) =
		set(index, value)

@Contract(pure = true)
operator fun <T> Queue<T>.get(index: Int) =
		get(index)

@Contract(pure = true)
operator fun <T> Queue<T>.set(index: Int, value: T) =
		set(index, value)

@Contract(pure = true)
operator fun <T> Stack<T>.get(index: Int) =
		get(index)

@Contract(pure = true)
operator fun <T> Stack<T>.set(index: Int, value: T) =
		set(index, value)



